File: FbxExporter/Assets/UTJ/FbxExporter/Scripts/FbxExporter.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace UTJ.FbxExporter
{
    public partial class FbxExporter
    {
        ExportOptions m_opt = ExportOptions.defaultValue;
        Context m_ctx;
        Dictionary<Transform, Node> m_nodes;

        public FbxExporter(ExportOptions opt)
        {
            m_opt = opt;
        }

        ~FbxExporter()
        {
            Release();
        }

        public void Release()
        {
            fbxeReleaseContext(m_ctx);
            m_ctx = Context.Null;
        }

        public bool CreateScene(string name)
        {
            Release();
            if (!m_ctx)
                m_ctx = fbxeCreateContext(ref m_opt);
            m_nodes = new Dictionary<Transform, Node>();
            return fbxeCreateScene(m_ctx, name);
        }

        public void AddNode(GameObject go)
        {
            if (go)
                FindOrCreateNodeTree(go.GetComponent<Transform>(), ProcessNode);
        }

        public bool Write(string path, Format format)
        {
            return fbxeWrite(m_ctx, path, format);
        }


        #region impl
        void ProcessNode(Transform trans, Node node)
        {
            var mr = trans.GetComponent<MeshRenderer>();
            var smr = trans.GetComponent<SkinnedMeshRenderer>();
            var terrain = trans.GetComponent<Terrain>();

            if (terrain)
                AddTerrain(node, terrain);
            else if (smr)
                AddSkinnedMesh(node, smr);
            else if (mr)
                AddMesh(node, mr);
        }

        Node FindOrCreateNodeTree(Transform trans, Action<Transform, Node> act)
        {
            if (!trans) { return Node.Null; }

            if (m_nodes.ContainsKey(trans))
            {
                return m_nodes[trans];
            }
            else
            {
                var parent = !trans.parent ? fbxeGetRootNode(m_ctx) : FindOrCreateNodeTree(trans.parent, act);
                var node = fbxeCreateNode(m_ctx, parent, trans.name);
                if (m_opt.transform)
                    fbxeSetTRS(m_ctx, node, trans.localPosition, trans.localRotation, trans.localScale);
                m_nodes.Add(trans, node);

                if (act != null) { act.Invoke(trans, node); }
                return node;
            }
        }

        bool AddMesh(Node node, Mesh mesh)
        {
            if (!mesh || mesh.vertexCount == 0) { return false; }
            if (!mesh.isReadable)
            {
                Debug.LogWarning("Mesh " + mesh.name + " is not readable and be ignored.");
                return false;
            }

            Topology topology = Topology.Triangles;

            var indices = new PinnedArray<int>(mesh.triangles);
            var points = new PinnedArray<Vector3>(mesh.vertices);
            var normals = new PinnedArray<Vector3>(mesh.normals); if (normals.Length == 0) normals = null;
            var tangents = new PinnedArray<Vector4>(mesh.tangents); if (tangents.Length == 0) tangents = null;
            var uv = new PinnedArray<Vector2>(mesh.uv); if (uv.Length == 0) uv = null;
            var colors = new PinnedArray<Color>(mesh.colors); if (colors.Length == 0) colors = null;
            fbxeAddMesh(m_ctx, node, points.Length, points, normals, tangents, uv, colors);
            fbxeAddMeshSubmesh(m_ctx, node, topology, indices.Length, indices, -1);

            int blendshapeCount = mesh.blendShapeCount;
            if (blendshapeCount > 0)
            {
                var deltaVertices = new PinnedArray<Vector3>(mesh.vertexCount);
                var deltaNormals = new PinnedArray<Vector3>(mesh.vertexCount);
                var deltaTangents = new PinnedArray<Vector3>(mesh.vertexCount);
                for (int bi = 0; bi < blendshapeCount; ++bi)
                {
                    string name = mesh.GetBlendShapeName(bi);
                    int frameCount = mesh.GetBlendShapeFrameCount(bi);
                    for (int fi = 0; fi < frameCount; ++fi)
                    {
                        float weight = mesh.GetBlendShapeFrameWeight(bi, fi);
                        mesh.GetBlendShapeFrameVertices(bi, fi, deltaVertices, deltaNormals, deltaTangents);
                        fbxeAddMeshBlendShape(m_ctx, node, name, weight, deltaVertices, deltaNormals, deltaTangents);
                    }
                }
            }

            return true;
        }

        bool AddMesh(Node node, MeshRenderer mr)
        {
            var mf = mr.gameObject.GetComponent<MeshFilter>();
            if (!mf)
                return false;
            return AddMesh(node, mf.sharedMesh);
        }

        bool AddSkinnedMesh(Node node, SkinnedMeshRenderer smr)
        {
            var mesh = smr.sharedMesh;
            if (!AddMesh(node, mesh))
                return false;

            var bones = smr.bones;
            var boneNodes = new PinnedArray<Node>(bones.Length);
            for (int bi = 0; bi < bones.Length; ++bi)
                boneNodes[bi] = FindOrCreateNodeTree(bones[bi], ProcessNode);
            var boneWeights = new PinnedArray<BoneWeight>(mesh.boneWeights);
            var bindposes = new PinnedArray<Matrix4x4>(mesh.bindposes);
            fbxeAddMeshSkin(m_ctx, node, boneWeights, boneNodes.Length, boneNodes, bindposes);
            return true;
        }

        bool AddTerrain(Node node, Terrain terrain)
        {
            var tdata = terrain.terrainData;
            var w = tdata.heightmapWidth;
            var h = tdata.heightmapHeight;
            var heightmap = tdata.GetHeights(0, 0, w, h);

            int vertexCount = w * h;
            int indexCount = (w - 1) * (h - 1) * 2 * 3;
            var vertices = new PinnedArray<Vector3>(vertexCount);
            var normals = new PinnedArray<Vector3>(vertexCount);
            var uv = new PinnedArray<Vector2>(vertexCount);
            var indices = new PinnedArray<int>(indexCount);
            fbxeGenerateTerrainMesh(heightmap, w, h, tdata.size,
                vertices, normals, uv, indices);

            Topology topology = Topology.Triangles;
            fbxeAddMesh(m_ctx, node, vertices.Length, vertices, normals, IntPtr.Zero, uv, IntPtr.Zero);
            fbxeAddMeshSubmesh(m_ctx, node, topology, indices.Length, indices, -1);

            return true;
        }
        #endregion
    }

}

File: Plugin/FbxExporter/FbxExporter.cpp
#include "pch.h"
#include "MeshUtils/MeshUtils.h"
#include "FbxExporter.h"
#include "fbxeContext.h"

using namespace fbxe;

fbxeAPI fbxe::IContext* fbxeCreateContext(const fbxe::ExportOptions *opt)
{
    return fbxe::CreateContext(opt);
}

fbxeAPI void fbxeReleaseContext(fbxe::IContext *ctx)
{
    if (!ctx) { return; }
    ctx->release();
}

fbxeAPI int fbxeCreateScene(fbxe::IContext *ctx, const char *name)
{
    if (!ctx) { return false; }
    return ctx->createScene(name);
}

fbxeAPI int fbxeWriteAsync(fbxe::IContext *ctx, const char *path, fbxe::Format format)
{
    if (!ctx) { return false; }
    return ctx->writeAsync(path, format);
}

fbxeAPI int fbxeIsFinished(fbxe::IContext *ctx)
{
    if (!ctx) { return false; }
    return ctx->isFinished();
}

fbxeAPI fbxe::Node* fbxeGetRootNode(fbxe::IContext *ctx)
{
    if (!ctx) { return nullptr; }
    return ctx->getRootNode();
}
fbxeAPI fbxe::Node* fbxeFindNodeByName(fbxe::IContext *ctx, const char *name)
{
    if (!ctx) { return nullptr; }
    return ctx->findNodeByName(name);
}

fbxeAPI fbxe::Node* fbxeCreateNode(fbxe::IContext *ctx, fbxe::Node *parent, const char *name)
{
    if (!ctx) { return nullptr; }
    return ctx->createNode(parent, name);
}

fbxeAPI void fbxeSetTRS(fbxe::IContext *ctx, fbxe::Node *node, float3 t, quatf r, float3 s)
{
    if (!ctx) { return; }
    ctx->setTRS(node, t, r, s);
}

fbxeAPI void fbxeAddMesh(fbxe::IContext *ctx, fbxe::Node *node, int num_vertices,
    const float3 points[], const float3 normals[], const float4 tangents[], const float2 uv[], const float4 colors[])
{
    if (!ctx) { return; }
    ctx->addMesh(node, num_vertices, points, normals, tangents, uv, colors);
}

fbxeAPI void fbxeAddMeshSubmesh(fbxe::IContext *ctx, fbxe::Node *node, fbxe::Topology topology, int num_indices, const int indices[], int material)
{
    if (!ctx) { return; }
    ctx->addMeshSubmesh(node, topology, num_indices, indices, material);
}

fbxeAPI void fbxeAddMeshSkin(fbxe::IContext *ctx, fbxe::Node *node, Weights4 weights[], int num_bones, fbxe::Node *bones[], float4x4 bindposes[])
{
    if (!ctx) { return; }
    ctx->addMeshSkin(node, weights, num_bones, bones, bindposes);
}

fbxeAPI void fbxeAddMeshBlendShape(fbxe::IContext *ctx, fbxe::Node *node, const char *name, float weight,
    const fbxe::float3 delta_points[], const fbxe::float3 delta_normals[], const fbxe::float3 delta_tangents[])
{
    if (!ctx) { return; }
    ctx->addMeshBlendShape(node, name, weight, delta_points, delta_normals, delta_tangents);
}



fbxeAPI void fbxeGenerateTerrainMesh(
    const float heightmap[], int width, int height, float3 size,
    float3 dst_vertices[], float3 dst_normals[], float2 dst_uv[], int dst_indices[])
{
    int num_vertices = width * height;
    int num_triangles = (width - 1) * (height - 1) * 2;
    auto size_unit = float3{ 1.0f / (width - 1), 1.0f, 1.0f / (height - 1) } *size;
    auto uv_unit = float2{ 1.0f / (width - 1), 1.0f / (height - 1) };

    parallel_invoke(
        [&]() {
        for (int iy = 0; iy < height; ++iy) {
            for (int ix = 0; ix < width; ++ix) {
                int i = iy * width + ix;
                dst_vertices[i] = float3{ (float)ix, heightmap[i], (float)iy } *size_unit;
                dst_uv[i] = float2{ (float)ix, (float)iy } *uv_unit;
            }
        }
    },
        [&]() {
        for (int iy = 0; iy < height - 1; ++iy) {
            for (int ix = 0; ix < width - 1; ++ix) {
                int i6 = (iy * width + ix) * 6;
                dst_indices[i6 + 0] = width*iy + ix;
                dst_indices[i6 + 1] = width*(iy + 1) + ix;
                dst_indices[i6 + 2] = width*(iy + 1) + (ix + 1);

                dst_indices[i6 + 3] = width*iy + ix;
                dst_indices[i6 + 4] = width*(iy + 1) + (ix + 1);
                dst_indices[i6 + 5] = width*iy + (ix + 1);
            }
        }
    });

    GenerateNormalsTriangleIndexed(dst_normals, dst_vertices, dst_indices, num_triangles, num_vertices);
}
